test: Add conformance tests

danieljbruce opened the pull request:
Add conformance tests that use the test proxy to ensure the client is running correctly. This PR should be very similar to [https://github.com/googleapis/java-bigtable/pull/1699](https://www.google.com/url?q=https://github.com/googleapis/java-bigtable/pull/1699&sa=D&source=buganizer&usg=AOvVaw3uY2epO8IEddjTnUwg00ez) which adds conformance tests to java-bigtable. A list of known test failures is also provided so that we only test against the test proxy tests that pass and our continuous integration pipeline passes. The idea is that we want to reduce this list over time as the test proxy improves.

danieljbruce committed:
Add conformance tests

Add conformance tests that use the test proxy to ensure the client is running correctly.

danieljbruce committed:
try this and see what it does

danieljbruce committed:
Revert "try this and see what it does"

This reverts commit aa18af61d6536769551b7625f1150dff8ccd9c5e.

danieljbruce committed:
Make the yaml config run node and not java

If the yaml config sets up node then the node test proxy can be tested.

danieljbruce committed:
Add echo to test

danieljbruce committed:
Add conformance to main CI pipeline instead

danieljbruce committed:
Add execution permissions

danieljbruce committed:
typo correction

danieljbruce committed:
Revert "typo correction"

This reverts commit 2bb0eed7223159ca1eb7ce32a91dcd210ea0bbda.

danieljbruce committed:
Revert "Add execution permissions"

This reverts commit 80d556962d5f723f06533cd2c2abcbf4b46ed0e3.

danieljbruce committed:
Revert "Add conformance to main CI pipeline instead"

This reverts commit 359d413e1315cb3727805282453926c6217f13bc.

danieljbruce committed:
After reverts, rearrange files for best setup

Add execution permissions to the conformance shell file.

danieljbruce committed:
Add logs to see what it does

danieljbruce committed:
Try adding more logs

danieljbruce committed:
Make sure it doesn’t get stuck

Add &

danieljbruce committed:
Correct conformance paths test run path

The path for the known failures is wrong and needs to be corrected.

danieljbruce committed:
Try without setting the proxy address

danieljbruce committed:
Fix path to the known failures again

danieljbruce committed:
Add a known failure so test will work

Errors out because skip has no argument

danieljbruce committed:
Set the proxy address

The proxy address needs to be set to where the test proxy is listening

danieljbruce committed:
Add known failures based on tests that fail

The known failures should be added to the file so that those tests can be skipped

danieljbruce committed:
Put all the test names on one line

By putting all the test names on one line maybe they will all be picked up to be skipped rather than just one of them.

danieljbruce committed:
Try using an ampersand to separate parameters

danieljbruce committed:
Revert "Try using an ampersand to separate parameters"

This reverts commit cb9f89d9ce0ced16ac88e4dad1e69abe9fabd33c.

danieljbruce committed:
Try unit array syntax

Using the unix array syntax should make it easier to process the files.

danieljbruce committed:
Revert "Try unit array syntax"

This reverts commit 1bb86d29c852800a2cc1b9b8358889aa288842d5.

danieljbruce committed:
Try reordering the items in known failures

See if output mutes first test

danieljbruce committed:
Try adding a comma

See if multiple arguments are accepted when a comma is added.

danieljbruce committed:
Use correct delimiter syntax for known failures

This commit uses the right syntax for including all tests that need to be skipped.

danieljbruce committed:
Add additional failures that were skipped before

Add tests to be skipped so that the conformance Github action passes.

danieljbruce committed:
Add the last known failure

One more known failure needs to be included

danieljbruce committed:
Remove output log location

The mentioned script is a leftover copy/paste error

danieljbruce committed:
Remove unecessary logging from script

Logging used to debug should not be part of the script anymore.

danieljbruce committed:
run the linter

Run the linter

danieljbruce committed:
remove console logs

danieljbruce committed:
Merge branch 'main' into conformance-tests

daniel-sanche commented on .github/workflows/conformance.yaml:
```diff
@@ -0,0 +1,41 @@
+  conformance:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
```
Is this needed? Looks like a duplicate of the step below

danieljbruce replied:
I bet it's not needed, but I'm not completely sure. The PR in Java has this twice so I figured there must be a reason it is there https://github.com/googleapis/java-bigtable/pull/1699/files. I'll investigate further.

danieljbruce replied:
I've been informed by @liujiongxin this check is probably for the Java client repo itself.

daniel-sanche replied:
Right, that makes sense. I forgot there were two repos involved!

daniel-sanche commented on .github/workflows/conformance.yaml:
```diff
@@ -0,0 +1,41 @@
+        path: cloud-bigtable-clients-test
+    - uses: actions/setup-node@v3
+      with:
+        node-version: 14
```
Do you think it would be useful to test with multiple node versions?

danieljbruce replied:
Yes

daniel-sanche commented on testproxy/known_failures.txt:
```diff
@@ -0,0 +1 @@
+TestCheckAndMutateRow_NoRetry_TransientError\|TestCheckAndMutateRow_Generic_Headers\|TestCheckAndMutateRow_Generic_DeadlineExceeded\|TestMutateRow_Generic_Headers\|TestMutateRow_Generic_DeadlineExceeded\|TestMutateRows_Generic_CloseClient\|TestReadModifyWriteRow_Generic_Headers\|TestReadModifyWriteRow_NoRetry_TransientError\|TestReadModifyWriteRow_Generic_DeadlineExceeded\|TestReadRow_Generic_Dea ...
```
I'm a little unsure if we should be skipping all these tests, maybe we should just let the test stay red until these are fixed?

danieljbruce replied:
Let's look into this further.

triplequark replied:
We should have two test paths: default and all tests enabled.

danieljbruce reviewed (commented):
Answer to review comments.

danieljbruce committed:
Test on multiple versions of node

The conformance tests may catch more issues to do with streaming and such if more versions of node are covered.

daniel-sanche commented on .kokoro/conformance.sh:
```diff
@@ -0,0 +1,44 @@
+cd $(dirname $0)/..
+
+# Build and start the proxy in a separate process
+pushd .
```
The pushd and popd shouldn't be necessary if you stay at the root

danieljbruce replied:
What exactly would you suggest here if we eliminate pushd and popd? I tried subshells locally https://unix.stackexchange.com/questions/13802/execute-a-specific-command-in-a-given-directory-without-cding-to-it, but the command hangs.

Looking at this further, I think in the Java example we wanted to go back to the root because we wanted to run https://github.com/googleapis/java-bigtable/pull/1699/files#diff-08b007a16e10891ada78edba8173c7323e74500fa6e1600c5e1dc7a721df5388R62. Since we don't need to run anything else in this script after the golang tests however, I think we can eliminate `pushd .` and `popd` and not worry about staying at the root.

daniel-sanche replied:
pushd and popd are only needed if you're changing directories. In the case of this specific block , it seems like a no-op, so yeah it makes sense to remove it

Alternatively, you could replace all the `cd` statements with `pushd` statements, which can be a lot cleaner if you have to move around the filesystem. But since this script is pretty simple, you can probably just stick with the `cds` you have

daniel-sanche commented on .kokoro/conformance.sh:
```diff
@@ -0,0 +1,44 @@
+pushd .
+pwd
+ls
+cd cloud-bigtable-clients-test/tests
```
shouldn't this be in pushd?

danieljbruce replied:
This code is copied straight from the https://github.com/googleapis/java-bigtable/pull/1699/files#diff-08b007a16e10891ada78edba8173c7323e74500fa6e1600c5e1dc7a721df5388R51-R56. Do you mean `pushd ./cloud-bigtable-clients-test/tests`? here and then eliminating the `pushd .` line?

That seems like a good idea that would condense the code, sure.

danieljbruce replied:
Now that  `pushd .` has been removed, no further action is required here.

daniel-sanche commented on .kokoro/conformance.sh:
```diff
@@ -0,0 +1,44 @@
+# Run the conformance test
+pushd .
+pwd
+ls
```
was pwd and ls added for debugging? Can probably remove them

danieljbruce replied:
Yes. These debugging commands have been removed now.

danieljbruce committed:
Eliminate logging for debugging

pwd and ls should not be left in the script

danieljbruce committed:
Remove pushd and popd

These two commands are unnecessary as nothing is run after them.

bhshkh commented on .kokoro/conformance.sh:
```diff
@@ -0,0 +1,40 @@
+RETURN_CODE=$?
+
+# Stop the proxy
+kill $proxyPID
```
This should be executed irrespective of when / how script exits. Exit traps can be used here. E.g. https://github.com/googleapis/google-cloud-go/blob/5de394be14cf852b99dde65fbeba3d02714f6845/storage/emulator_test.sh#L64-L70

danieljbruce replied:
Sure. Makes sense. So would you suggest adding `trap kill $proxyPID` somewhere at the top of the file and removing this line? In that case would any exit signal still get propogated up to the user?

bhshkh replied:
Add a cleanup function.
Trap on EXIT signal which would catch any exit signal.

Similar to https://github.com/googleapis/google-cloud-go/blob/5de394be14cf852b99dde65fbeba3d02714f6845/storage/emulator_test.sh#L70

danieljbruce replied:
Done

daniel-sanche replied:
Was the trap change pushed? I don't see the changes here

danieljbruce replied:
In my git client at about 1:30 I made the changes and pushed them. Not sure why they don't show up.

![image](https://github.com/googleapis/nodejs-bigtable/assets/8935272/5a547bff-d789-4de5-8d8c-f90d8ce79a0b)

danieljbruce replied:
It is called "Add a cleanup operation"

danieljbruce committed:
Remove known failures list of test

As required, remove the list of known failures

danieljbruce committed:
Add a cleanup operation

The cleanup operation should happen no matter how the program exits.

danieljbruce committed:
Empty-Commit
